File: manual/code/test_suites/cpp/FunctionBuildTest/tests/test_Whoosh_index_writer.cpp
#include <gtest/gtest.h>
#include <string>
#include <filesystem>
#include "../src/function_Whoosh_index_writer.cpp"

class TestManageSearchDocuments : public ::testing::Test {
protected:
    std::string index_dir = "test_index";
    WhooshIndexManager manager;

    void SetUp() override {
        cleanup();
        initialize_index(index_dir);
    }

    void TearDown() override {
        cleanup();
    }

    void cleanup() {
        if (std::filesystem::exists(index_dir)) {
            std::filesystem::remove_all(index_dir);
        }
    }

    void initialize_index(const std::string& index_dir) {
        manager.manage_and_search_documents<std::string>(index_dir, "Document 1: Test text.");
    }
};

TEST_F(TestManageSearchDocuments, test_add_document) {
    std::string result = manager.manage_and_search_documents<std::string>(
        index_dir, "Document 2: Another test text.", "Document");
    EXPECT_TRUE(result.find("Document 2: Another test text.") != std::string::npos);
}

TEST_F(TestManageSearchDocuments, test_empty_index_search) {
    std::string result = manager.manage_and_search_documents<std::string>(
        index_dir, "", "Nonexistent");
    EXPECT_EQ(result, "No documents found");
}

TEST_F(TestManageSearchDocuments, test_empty_index) {
    std::string result = manager.manage_and_search_documents<std::string>(
        index_dir, "Document 2: Another test text.", "Document");
    EXPECT_TRUE(result.find("Document 2: Another test text.") != std::string::npos);
}

TEST_F(TestManageSearchDocuments, test_multiple_updates_on_same_document) {
    manager.manage_and_search_documents<std::string>(
        index_dir, "Updated Document: Final version.");
    std::string result = manager.manage_and_search_documents<std::string>(
        index_dir, "", "Updated");
    EXPECT_TRUE(result.find("Updated Document: Final version.") != std::string::npos);
}

TEST_F(TestManageSearchDocuments, test_search_specific_document) {
    std::string result = manager.manage_and_search_documents<std::string>(
        index_dir, "", "Test");
    EXPECT_TRUE(result.find("Document 1: Test text.") != std::string::npos);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: data/python/tests/test_Whoosh_index_writer.py
import os
import shutil
import unittest
from function_Whoosh_index_writer import WhooshIndexManager

class TestManageSearchDocuments(unittest.TestCase):
    def setUp(self):
        self.index_dir = "test_index"
        self.manager = WhooshIndexManager()
        self._initialize_index(self.index_dir)

    def tearDown(self):
        if os.path.exists(self.index_dir):
            shutil.rmtree(self.index_dir)

    def _initialize_index(self, index_dir):
        self.manager.manage_and_search_documents(index_dir, document_text="Document 1: Test text.")

    def test_add_document(self):
        result = self.manager.manage_and_search_documents(self.index_dir, document_text="Document 2: Another test text.", search_query="Document")
        self.assertIn("Document 2: Another test text.", result)

    def test_empty_index_search(self):
        result = self.manager.manage_and_search_documents(self.index_dir, search_query="Nonexistent")
        self.assertEqual(result, "No documents found")

    def test_empty_index(self):
        result = self.manager.manage_and_search_documents(self.index_dir, document_text="Document 2: Another test text.", search_query="Document")
        self.assertIn("Document 2: Another test text.", result)

    def test_multiple_updates_on_same_document(self):
        self.manager.manage_and_search_documents(self.index_dir, document_text="Updated Document: Final version.")
        result = self.manager.manage_and_search_documents(self.index_dir, search_query="Updated")
        self.assertIn("Updated Document: Final version.", result)

    def test_search_specific_document(self):
        result = self.manager.manage_and_search_documents(self.index_dir, search_query="Test")
        self.assertIn("Document 1: Test text.", result)

if __name__ == "__main__":
    unittest.main()
